core/src/main/java/org/apache/iceberg/BaseEntriesTable.java

```diff
@@ -125,31 +130,107 @@ ManifestFile manifest() {
 
     @Override
     public CloseableIterable<StructLike> rows() {
-      // Project data-file fields
-      CloseableIterable<StructLike> prunedRows;
-      if (manifest.content() == ManifestContent.DATA) {
-        prunedRows =
+      Types.NestedField readableMetricsField = projection.findField(MetricsUtil.READABLE_METRICS);
+
+      if (readableMetricsField == null) {
+        CloseableIterable<StructLike> entryAsStruct =
             CloseableIterable.transform(
-                ManifestFiles.read(manifest, io).project(fileSchema).entries(),
-                file -> (GenericManifestEntry<DataFile>) file);
+                entries(fileSchema),
+                entry -> (GenericManifestEntry<? extends ContentFile<?>>) entry);
+
+        StructProjection structProjection = projectNonReadable(projection);
+        return CloseableIterable.transform(entryAsStruct, structProjection::wrap);
       } else {
-        prunedRows =
-            CloseableIterable.transform(
-                ManifestFiles.readDeleteManifest(manifest, io, specsById)
-                    .project(fileSchema)
-                    .entries(),
-                file -> (GenericManifestEntry<DeleteFile>) file);
+        Set<Integer> readableMetricsIds = TypeUtil.getProjectedIds(readableMetricsField.type());
```

Comment:
I had some idea to make this method even a bit more clear, again by making some private method that better describe these steps:

```
    @Override
    public CloseableIterable<StructLike> rows() {
      Types.NestedField readableMetricsField = projection.findField(MetricsUtil.READABLE_METRICS);

      if (readableMetricsField == null) {
        CloseableIterable<StructLike> entryAsStruct =
            CloseableIterable.transform(
                entries(fileProjection),
                entry -> (GenericManifestEntry<? extends ContentFile<?>>) entry);

        StructProjection structProjection = projectNonReadable(projection);
        return CloseableIterable.transform(entryAsStruct, structProjection::wrap);
      } else {
        Schema requiredFileProjection = requiredFileProjection(fileProjection);
        Schema actualProjection = removeReadableMetrics(projection, readableMetricsField);

        StructProjection structProjection = projectNonReadable(actualProjection);
        return CloseableIterable.transform(
            entries(requiredFileProjection),
            entry ->
                withReadableMetrics(structProjection, entry, readableMetricsField));
      }
    }
```


Now its a bit more symmetrical.  

We would need following methods:
```
    /**
     * Remove virtual columns from the file projection and ensure that the underlying metrics
     * used to create those columns are part of the file projection
     * @return file projection with required columns to read readable metrics
     */
    private Schema requiredFileProjection(Schema fileProjection) {
      Schema projectionForReadableMetrics =
          new Schema(
              MetricsUtil.READABLE_METRIC_COLS.stream()
                  .map(MetricsUtil.ReadableMetricColDefinition::originalCol)
                  .collect(Collectors.toList()));
      return TypeUtil.join(fileProjection, projectionForReadableMetrics);
    }

    private Schema removeReadableMetrics(Schema projection, Types.NestedField readableMetricsField) {
      Set<Integer> readableMetricsIds = TypeUtil.getProjectedIds(readableMetricsField.type());
      return TypeUtil.selectNot(projection, readableMetricsIds);
    }
```

I also change withReadableMetrics a bit for this to work:

```
    private StructLike withReadableMetrics(
        StructProjection structProjection,
        ManifestEntry<? extends ContentFile<?>> entry,
        Types.NestedField readableMetricsField) {
      int projectionColumnCount = projection.columns().size();
      int metricsPosition = projection.columns().indexOf(readableMetricsField);

      StructProjection entryStruct = structProjection.wrap((StructLike) entry);

      StructType projectedMetricType =
          projection.findField(MetricsUtil.READABLE_METRICS).type().asStructType();
      MetricsUtil.ReadableMetricsStruct readableMetrics =
          MetricsUtil.readableMetricsStruct(dataTableSchema, entry.file(), projectedMetricType);
      
      return new ManifestEntryStructWithMetrics(
          projectionColumnCount, metricsPosition, entryStruct, readableMetrics);
    }
```

Reply:
thank you @szehon-ho . that does look much easier to follow the logic. I changed a bit since checkStyle fail on some method variable shadow class variable but mostly followed your approach and appreciate your help!